scripts/build_ngx_pagespeed.sh

```diff
@@ -0,0 +1,331 @@
+#!/bin/bash
+
+# Put everything in one big function so that if this script is executed after a
+# partial download it doesn't do anything.
+function build_ngx_pagespeed() {
+  getopt --test
+  if [ "$?" != 4 ]; then
+    echo "Your version of getopt is too old.  Exiting with no changes made."
+    # Even Centos 5 and Ubuntu 10 LTS have new-style getopt, so I don't expect
+    # this to be hit in practice on systems that are actually able to run
+    # PageSpeed.
+    exit 1
+  fi
+
+  # If we set -e or -u then users of this script will see it silently exit on
+  # failure.  Instead we need to check the exit status of each command
+  # manually.  The run function (see below) handles exit-status checking for
+  # system-changing commands.
+
+  function usage() {
```

Comment:
Is there any reason you need to nest all your functions within "main"? I think it would be more readable if they were declared at global scope.